[Export] Add one-click deploy over SSH for the desktop exports.

bruvzg opened the pull request:
Adds one-click deploy and remote debug option for desktop targets, similar to the same feature for Android and Web.

- Add ZIP export option for Linux and Windows.
- Add one-click deploy over SSH for Linux, macOS and Windows targets.

https://user-images.githubusercontent.com/7645683/180428609-4c7b5986-0ecf-49e1-b8e3-b5794712cb45.mov

TODO:
- [x] Support for Linux targets.
- [x] Support for macOS targets (the same as Linux, with exception for bundle run command).
- [x] Support for Windows targets (Windows 10+ should have SSH server as a standard OS component, but `mktemp` and unzip should be replaced with PowerShell equivalents). The built-in version of OpenSSH is running as a service and cannot access a desktop session, therefor task scheduler is used to start / terminate the uploaded project.
- [x] Cleanup temp files on remote after finishing debug session.
- [x] Add editor settings to set custom paths for `scp` and `ssh`.
- [x] Replace ugly 16x16 bitmap "run" icons with the scalable SVGs.

Related proposals https://github.com/godotengine/godot-proposals/issues/4897, https://github.com/godotengine/godot-proposals/issues/2608, not exactly the same but should simplify testing on the portables like Steam Deck, or multiple platforms in general.

Required tools / config to use this:
- SSH with non-interactive authentication (e.g., key based auth controlled by ssh-agent).
- For Unix targets: `kill`, `mktemp`, `rm`, `unzip` commands/utilities available and in path.
- For Linux, additional environment variables for desktop session might be needed (can set in export options), by default it's setting only `DISPLAY=:0`;
- For Windows: PowerShell installed (installed by default), target user must have active desktop session and have permission to use task scheduler (enabled by default for all users, but might be blocked via Group Policy).

bruvzg committed:
[Export] Add one-click deploy over SSH for the desktop exports.

Add one-click deploy over SSH for the desktop exports.
Add ZIP export option for Linux and Windows.
Change export plugin icons to SVG format.

Calinou commented:
I suppose this also addresses https://github.com/godotengine/godot-proposals/issues/1072 to an extent.

bruvzg commented:
> Feel free to use [platform-icons.zip](https://github.com/godotengine/godot/files/9169603/platform-icons.zip) (which I've made over 1 year ago, but never got to work within Godot itself).

Changed platform icons to this set, and run icons to the gray version of them (+ added UWP icon), also using SVG icons fixes icon scaling in the hiDPI mode.

| Before | After | 
|---|---|
| <img width="474" alt="Screenshot 2022-07-24 at 13 31 34" src="https://user-images.githubusercontent.com/7645683/180664359-65d96f28-8fb0-4257-af27-40a2398a0a11.png"> | <img width="474" alt="Screenshot 2022-07-24 at 14 03 22" src="https://user-images.githubusercontent.com/7645683/180664362-67b0906b-f1d6-4426-afbd-ebdbf4d520fd.png"> |
| <img width="397" alt="Screenshot 2022-07-24 at 13 31 24" src="https://user-images.githubusercontent.com/7645683/180664516-c57e1859-b8ab-4d57-ac4d-14295cffce9c.png"> | <img width="397" alt="Screenshot 2022-07-24 at 23 23 22" src="https://user-images.githubusercontent.com/7645683/180664534-577a8aca-f16a-44a7-a6fd-5aa4c75dab7c.png"> |

bruvzg commented:
> I suppose this also addresses https://github.com/godotengine/godot-proposals/issues/1072 to an extent.

<img width="388" alt="Screenshot 2022-07-27 at 13 42 04" src="https://user-images.githubusercontent.com/7645683/181228139-04e398f1-f809-46f5-9723-583f544c0c70.png">

With the full set of remote debug options enabled, toolbar become quite cluttered, so I guess some sort of UI redesign might be necessary. And running projects using local export templates might be desired, as well (you can use this PR for this by connecting to the localhost, but it's definitely not a proper solution).

aaronfranke commented:
I've already commented in https://github.com/godotengine/godot-proposals/issues/2124, but I would rather use macOS Finder icon for macOS. The Apple logo is too generic, and it could be confused for iOS at a glance.

![](https://webclt.com/macos-flat-noborders.svg)

[macos-flat-noborders.zip](https://github.com/godotengine/godot-proposals/files/5858225/macos-flat-noborders.zip)

bruvzg commented:
> I've already commented in https://github.com/godotengine/godot-proposals/issues/2124, but I would rather use macOS Finder icon for macOS. The Apple logo is too generic, and it could be confused for iOS at a glance.

Changed it to Finder icon. Not really sure monochrome run icon ended up good, but it's probably better, at least for consistency with previous versions.

<img width="249" alt="Screenshot 2022-07-28 at 22 12 20" src="https://user-images.githubusercontent.com/7645683/181691188-90f43032-5985-4f5c-97d6-474c4931ecb5.png">

reduz commented on platform/linuxbsd/export/export_plugin.cpp:
```diff
@@ -200,3 +265,185 @@ Error EditorExportPlatformLinuxBSD::fixup_embedded_pck(const String &p_path, int
 	return OK;
 }
+
+Ref<Texture2D> EditorExportPlatformLinuxBSD::get_run_icon() const {
```
I feel a standalone file should be added for this export platform given its getting a bit big

bruvzg replied:
Like a separate files for `export` and `run` parts?

akien-mga replied:
I think @reduz might have misread the file paths. This is already a dedicated file (but indeed it used to be that both LinuxBSD and Windows were together in `editor/editor_export.cpp`).

reduz commented on platform/macos/export/export_plugin.cpp:
```diff
@@ -1678,9 +1574,191 @@ bool EditorExportPlatformMacOS::can_export(const Ref<EditorExportPreset> &p_pres
 }
 
-EditorExportPlatformMacOS::~EditorExportPlatformMacOS() {
+bool EditorExportPlatformMacOS::poll_export() {
```
Likewise I think this platform should get its own file

reduz reviewed (dismissed):
It seems fairly good. Do you know if sending only the binary and obtaining files via LAN is supported? I don't see the code for this, but maybe new code is not needed.

bruvzg commented:
> Do you know if sending only the binary and obtaining files via LAN is supported? I don't see the code for this, but maybe new code is not needed.

Should need a little modification to the export code to skip packing files, the rest should work as is, but I have not tested it.

bruvzg commented:
> >    Do you know if sending only the binary and obtaining files via LAN is supported? I don't see the code for this, but maybe new code is not needed.
>
> Should need a little modification to the export code to skip packing files, the rest should work as is, but I have not tested it.

Simply adding export code to skip packing files doesn't work.

It's exporting without PCK (or with empty PCK, I have tried both), uploading it, running remote executable with `--remote-fs` argument, and starting file server but nothing happens. The same is true if I try to run it locally.

So it's either something wrong with Godot file server or desktop platforms are missing something to work with `--remote-fs`.

akien-mga reviewed (commented):
Some initial comments, didn't do a thorough review.
Did some testing though, will add further comments in the PR.

akien-mga commented:
I tested this PR to do SSH export to a Steam Deck, it's working!

I ran into a few UX issues and there's more work that might be needed either here, or in a follow-up, or documentation, to make the process deploying to Steam Deck more convenient.

Some general notes:

- Editing the SSH and SCP args as a `PackedStringArray` is not so convenient. I think it might be better to expose those settings as a multiline text field and do the `split(" ")` to get a `PackedStringArray` in editor code.

- Here's the config I'm using to export to Steam Deck. It's copied from what the "Steam Devkit Management Tool" does (which is open source btw https://gitlab.steamos.cloud/devkit/steamos-devkit).

![image](https://user-images.githubusercontent.com/4701338/194527639-ade8de31-a90b-47ae-831f-5d150bb50b3a.png)

Aside from the host, and actually expanding `$HOME` instead of hardcoding it, this should be the common config for Steam Deck export for all users who used the SteamOS devkit once to get the RSA key in their home.

- SteamOS can run in two modes: SteamUI / gamescope which is the "console" like experience, and desktop mode which is a KDE Plasma on Wayland environment. I tried both and had different issues on each (CC @TTimo if you're interested *and* available).

  * In console mode, the export worked fine and I could see in my local terminal that the game is running (and remotely via SSH also in `top`), however the game didn't show up on screen. I'll have to dig a bit into Steam Deck internals and see if there's an easy way to force running a process in console mode. Otherwise, we'd need to write an export plugin that replicates some of the config that the SteamOS devkit does to "install" the temporary export as a non-Steam game and run it with `steam://run/...`. I was hoping that the SteamOS devkit would actually have an "Install and auto-start" I could reference but it doesn't seem to (would be a useful feature IMO *wink wink* Timothee).

  * In desktop mode, the export also worked fine but Godot actually failed to run, with this error (don't mind the 4.0.beta1 version, I just reused a template I had available with my custom editor build from this PR):

```
Godot Engine v4.0.beta1.official.20d667284 - https://godotengine.org
Authorization required, but no authorization protocol specified
ERROR: X11 Display is not available
   at: DisplayServerX11 (platform/linuxbsd/display_server_x11.cpp:4681)
Authorization required, but no authorization protocol specified

(zenity:3896): Gtk-WARNING **: 11:21:04.407: cannot open display: :0
ERROR: Unable to create DisplayServer, all display drivers failed.
   at: setup2 (main/main.cpp:1846)
```

AFAIK SteamOS's KDE runs on Wayland (initiated by SDDM), and it seems not to have `~/.Xauthority` readily available. From the Steam Deck desktop, navigation to the `/tmp/tmp_*` folder where it gets copied to run the executable works fine. But our SSH / `xauth` magic doesn't seem to work properly for this Wayland setup.

https://unix.stackexchange.com/a/327683 helped me figure out a workaround (running on the Steam Deck via SSH):

```bash
export DISPLAY=:0
rm -f ~/.Xauthority
touch ~/.Xauthority
magic_file=$(ps aux | grep Xorg | grep auth | sed "s/.*-auth \([^ ]*\) .*/\1/")
magic=$(xauth -f "$magic_file" list | sed "s/.*\($DISPLAY.*\)/\1/")
xauth add $magic
```

Then the game autostart properly when deployed from Godot (in desktop mode only).
This workaround needs to be redone after a reboot/session login of the Deck.

There might be a better / more standard way to set up the right Xauthority of course (and also better command line hacks to do the above, I have 0 `awk` fu so I tend to abuse `sed` a lot ;)).

- Remember to enable "Deploy with Remote Debug" in the Debug menu. We might want to consider making it default? Then syncing scene changes works fine (both ways, you can edit the running game from the editor, and see the state in the Remote scene tree). Syncing script changes seems a bit broken but that's unrelated to this PR, it must be a pre-existing issue.

bruvzg commented:
Changed args to the multiline text, also added script templates as export options, instead of executing hardcoded commands (I guess it might be worth to add preset selector for it as well).

![Screenshot 2022-10-21 at 09 19 55](https://user-images.githubusercontent.com/7645683/197129151-a0a576f1-653b-4c7b-a878-8747c68b0f02.png)

BAndiT1983 commented:
Works great on Steam Deck! Stepping through the script code in debug sessions is working fine too.

- Is this feature still "work in progress"? No updates for a while, so decided to ask
- Are there plans to enable the stop button? Currently it's disabled, at least in my build, so stopping the session is only possible through stop and uninstall
- Is it possible to add something like rsync, to avoid uninstalling of the files and to upload the diffs between sessions? Maybe there is already something and I've missed it

Could adjust some things too, but would be good to know the state first

bruvzg commented:
> Is this feature still "work in progress"? No updates for a while, so decided to ask

Oh, I completely forgot to mark is as ready after adding script templates to the export options. Rebased, and will re-test it to make sure everything is still working tomorrow.

> Are there plans to enable the stop button? Currently it's disabled, at least in my build, so stopping the session is only possible through stop and uninstall

It's the same for Android and Web, so probably no.

> Is it possible to add something like rsync, to avoid uninstalling of the files and to upload the diffs between sessions? Maybe there is already something and I've missed it

No, it's not supported. Should be easy to add, but since it's uploading the project as a single zip, it probably won't do much.

akien-mga reviewed (approved):
Looks great to me! I'll merge after beta 12 (releasing tomorrow) so we have a few days to test before the next beta.

akien-mga commented:
Thanks!

eh-jogos commented:
Is there any documentation on how to get this working? 

I work on a linux machine and would like to one click deploy to a windows device. I tried the latest one click deploy page, but it seems to be the same one as the 3.x stable version https://docs.godotengine.org/en/latest/tutorials/export/one-click_deploy.html

BlackDragonBE commented:
All this does for me on the Steam Deck is creating an empty directory in /tmp:

```
Creating temporary directory...
/tmp/tmp.AY7eTuMc16
```

This really needs some documentation.

BAndiT1983 commented:
Works for me on SteamDeck, was following the docs by @akien-mga and @bruvzg above, what were your steps to set the connection up?

BlackDragonBE commented:
@BAndiT1983 
Probbaly something I'm doing wrong then. Any help would be greatly appreciated!
Here's what I'm doing:

1. I've set up SSH, created a ssh key pair on the deck and copied the key over to my ~/.ssh folder. I've tested the connection in Powershell and I can connect via ssh without having to provide a password.
2. In my Godot project, I've added a Linux export, checked Enabled for SSH Remote Deploy and entered the Host.
3. I added this to Extra Args SSH and Extra Args SCP respectively:

```
-o StrictHostKeyChecking=no -o IdentitiesOnly=yes -t
```

```
-o StrictHostKeyChecking=no
```

4. On the Steam Deck I executed these commands via SSH:

```
export DISPLAY=:0
rm -f ~/.Xauthority
touch ~/.Xauthority
magic_file=$(ps aux | grep Xorg | grep auth | sed "s/.*-auth \([^ ]*\) .*/\1/")
magic=$(xauth -f "$magic_file" list | sed "s/.*\($DISPLAY.*\)/\1/")
xauth add $magic
```

5. I try to do a remote (with Debug > Remote Debug enabled), but all I get is this in my output:

```
Creating temporary directory...
/tmp/tmp.o7yKIuQ3C3
```

When I check on the deck, there is indeed a new tmp folder in /tmp, but it's empty.  
I thought it might be a SCP issue, but when I run SCP manually to copy a test zip, it works fine, just not via Godot.

Thanks!

BAndiT1983 commented:
Will cross-check when i get to my Deck, but overall it looks correct at first glance

BlackDragonBE commented:
I'm still getting the same issue as detailed in https://github.com/godotengine/godot/pull/63312#issuecomment-1435949729
A directory in /tmp is created and that's it. Can anyone else using Windows test if it works for them?

BAndiT1983 commented:
Sorry for long absence (work and other stuff came in-between), have finally checked under Windows and the export is also not starting automatically on Deck, tried to run the shell script from the /tmp/tmp..., but it fails, at least the files are transferred. Will analyse why it SIGSEGVs there. Does it run on your Deck?

BAndiT1983 commented:
Same workaround from `Edit 3` in the comment above is also valid for deployments from Windows. Am not getting any display error or similar, it just doesn't start after deploying files, but after using xhost line it starts just fine via remote deployment.

BlackDragonBE commented:
@BAndiT1983 
I've executed `xhost +si:localuser:$USER` on the deck and tried again, but all it does in Godot after trying remote deployement is this:

```
Creating temporary directory...
/tmp/tmp.4mx2BhRDMn
```

With --verbose:

```
Creating temporary directory...
Executing: ssh -p 22 -q -o LogLevel=error -o BatchMode=yes -o StrictHostKeyChecking=no deck@dragondeck mktemp -d
Exit code: 259, Output: /tmp/tmp.x1UKLSSUKn

/tmp/tmp.x1UKLSSUKn
```

Nothing happens on the deck.
Can you share your export settings?

BlackDragonBE commented:
Update: I was trying it on Godot 4.0.0, after updating to Godot 4.0.2 it works perfectly!

BAndiT1983 commented:
If it stalls with `Creating temporary directory...` then the connection settings are not correct, had this behaviour too, when the IP address or SSH settings were wrong.
